### PR TITLE
Add RandomCropGenerator

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

**New feature** 

## Description:

Extracts a random crop generator operator (before it could only be achieved via `decoders.image_random_crop` or `resized_random_crop`.
This enables the use case from https://github.com/NVIDIA/DALI/issues/2576

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [x] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [x] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by szkarpinski** [comment]
I'm afraid you also need to add checkpointing test for this operator (or coverage test will fail). Checkpointing tests are in [test/python/checkpointing/test_dali_checkpointing.py](https://github.com/NVIDIA/DALI/blob/main/dali/test/python/checkpointing/test_dali_checkpointing.py) and the test should look more or less like this:
```py
@random_signed_off("random_crop_generator")
def test_random_crop_generator():
    check_single_input_operator(fn.segmentation.random_object_bbox, "cpu", ...)
```
See other tests for random ops for reference.

File: dali/operators/image/crop/random_crop_generator.cc [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// limitations under the License.
+
+#include "dali/operators/image/crop/random_crop_generator.h"
+
+namespace dali {
+
+DALI_SCHEMA(RandomCropGenerator)
+  .DocStr(R"code(Produces a cropping window with a randomly selected area and aspect ratio.
+
+Expects a one-dimensional input representing the shape of the input we want to crop (HW or HWC representation).)code")
```
Comment (szkarpinski) [inline dali/operators/image/crop/random_crop_generator.cc:22]: It's probably worth mentioning that the output format is anchor+shape

**Comment by jantonguirao** [comment]
> I'm afraid you also need to add checkpointing test for this operator (or coverage test will fail). Checkpointing tests are in [test/python/checkpointing/test_dali_checkpointing.py](https://github.com/NVIDIA/DALI/blob/main/dali/test/python/checkpointing/test_dali_checkpointing.py) and the test should look more or less like this:
> 
> ```python
> @random_signed_off("random_crop_generator")
> def test_random_crop_generator():
>     check_single_input_operator(fn.random_crop_generator, "cpu", ...)
> ```
> 
> See other checkpointing tests for random ops for reference.

Done

File: dali/operators/image/crop/random_crop_generator.cc (hunk 1 above)
Comment (jantonguirao) [reply dali/operators/image/crop/random_crop_generator.cc]: Done

**Review by szkarpinski — APPROVED** [review]

**Review by klecki — COMMENTED** [review]
As a general question, do you envision a GPU variant for this operator? How would that impact the API?

File: dali/operators/image/crop/random_crop_generator.cc [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+#include "dali/operators/image/crop/random_crop_generator.h"
+
+namespace dali {
+
+DALI_SCHEMA(RandomCropGenerator)
+  .DocStr(R"code(Produces a cropping window with a randomly selected area and aspect ratio.
+
+Expects a one-dimensional input representing the shape of the input we want to crop (HW or HWC representation).
+
+Produces two outputs, representing the anchor and shape of the cropping window.)code")
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/operators/image/crop/random_crop_generator.cc:24]: `fn.crop`'s crop_pos* actually have a different meaning

**Comment by jantonguirao** [comment]
> As a general question, do you envision a GPU variant for this operator? How would that impact the API?

I think it does not make sense to add a GPU variant to it, since this is a metadata op

File: dali/operators/image/crop/random_crop_generator.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (klecki) [reply dali/operators/image/crop/random_crop_generator.cc:24]: Oh, they are also relative. Still, I would think about using `crop` with `random_crop_generator` and than would be looking for a way how to actually use those crops.

**Review by klecki — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**